nexus/client.go

```diff
@@ -0,0 +1,543 @@
+// wait period provided via [ExecuteOperationOptions] exceeds, in which case an [ErrOperationStillRunning] error is
+// returned.
+//
+// The wait time is capped to the deadline of the provided context. Make sure to handle both context deadline errors and
+// [ErrOperationStillRunning].
+//
+// Note that the wait period is enforced by the server and may not be respected if the server is misbehaving. Set the
+// context deadline to the max allowed wait period to ensure this call returns in a timely fashion.
+//
+// ⚠️ If this method completes successfully, the returned response's body must be read in its entirety and closed to
+// free up the underlying connection.
+func (c *Client) ExecuteOperation(ctx context.Context, request ExecuteOperationOptions) (*http.Response, error) {
+	result, err := c.StartOperation(ctx, request.intoStartOptions())
+	if err != nil {
+		return nil, err
+	}
+	if result.Successful != nil {
+		return result.Successful, nil
+	}
+	handle := result.Pending
+	return handle.GetResult(ctx, request.intoGetResultOptions())
+}
+
+// NewHandle gets a handle to an asynchronous operation by name and ID.
+// Does not incur a trip to the server.
+// Fails if provided an invalid operation or ID.
+func (c *Client) NewHandle(operation string, operationID string) (*OperationHandle, error) {
+	var es []error
+	if operation == "" {
+		es = append(es, errEmptyOperationName)
+	}
+	if operationID == "" {
+		es = append(es, errEmptyOperationID)
+	}
+	if len(es) > 0 {
+		return nil, errors.Join(es...)
+	}
+	return &OperationHandle{
+		client:    c,
+		Operation: operation,
+		ID:        operationID,
+	}, nil
+}
+
+func (c *Client) sendGetOperationRequest(ctx context.Context, request *http.Request) (*http.Response, error) {
+	response, err := c.options.HTTPCaller(request)
+	if err != nil {
+		return nil, err
+	}
+
+	if response.StatusCode == http.StatusOK {
+		return response, nil
+	}
+
+	// Do this once here and make sure it doesn't leak.
+	body, err := readAndReplaceBody(response)
+	if err != nil {
+		return nil, err
+	}
+
+	switch response.StatusCode {
+	case http.StatusRequestTimeout:
+		return nil, errOperationWaitTimeout
+	case statusOperationRunning:
+		return nil, ErrOperationStillRunning
+	case statusOperationFailed:
+		state, err := getUnsuccessfulStateFromHeader(response, body)
+		if err != nil {
+			return nil, err
+		}
+		failure, err := failureFromResponse(response, body)
+		if err != nil {
+			return nil, err
+		}
+		return nil, &UnsuccessfulOperationError{
+			State:   state,
+			Failure: failure,
+		}
+	default:
+		return nil, newUnexpectedResponseError(fmt.Sprintf("unexpected response status: %q", response.Status), response, body)
+	}
+}
+
+// NewCompletionHTTPRequest creates an HTTP request deliver an operation completion to a given URL.
+func NewCompletionHTTPRequest(ctx context.Context, url string, completion OperationCompletion) (*http.Request, error) {
```

Comment:
All of this completion stuff belongs somewhere else IMO. It clutters up the client file I think.

Reply:
Sure, I'll move it.